picard/browser/addrelease.py

```diff
@@ -0,0 +1,235 @@
+
+
+try:
+    import jwt
+    import jwt.exceptions
+except ImportError:
+    log.debug('PyJWT not available, addrelease functionality disabled')
+    jwt = None
+
+__key = token_bytes()  # Generating a new secret on each startup
+__algorithm = 'HS256'
+
+_form_template = '''<!doctype html>
+<meta charset="UTF-8">
+<html>
+<head>
+    <title>{title}</title>
+</head>
+<body>
+    <form action="{action}" method="post">
+        {form_data}
+        <input type="submit" value="{submit_label}">
+    </form>
+    <script>document.forms[0].submit()</script>
+</body>
+'''
+
+_form_input_template = '<input type="hidden" name="{name}" value="{value}" >'
+
+
+class InvalidTokenError(Exception):
+    pass
+
+
+class NotFoundError(Exception):
+    pass
+
+
+def is_available():
+    return jwt is not None
+
+
+def is_enabled():
+    tagger = QCoreApplication.instance()
+    return tagger.browser_integration.is_running
+
+
+def submit_cluster(cluster):
+    _open_url_with_token({'cluster': hash(cluster)})
+
+
+def submit_file(file):
+    _open_url_with_token({'file': file.filename})
+
+
+def serve_form(token):
+    try:
+        payload = jwt.decode(token, __key, algorithms=__algorithm)
+        log.debug('received JWT token %r', payload)
+        if 'cluster' in payload:
+            cluster = _find_cluster(payload['cluster'])
+            if not cluster:
+                raise NotFoundError('Cluster not found')
+            return _get_cluster_form(cluster)
+        elif 'file' in payload:
+            file = _find_file(payload['file'])
+            if not file:
+                raise NotFoundError('File not found')
+            return _get_file_form(file)
+        else:
+            raise InvalidTokenError
+    except jwt.exceptions.InvalidTokenError:
+        raise InvalidTokenError
+
+
+def extract_discnumber(metadata):
+    try:
+        discnumber = metadata.get('discnumber', '1').split('/')[0]
+        return int(discnumber)
+    except ValueError:
+        return 1
+
+
+def _open_url_with_token(payload):
+    token = jwt.encode(payload, __key, algorithm=__algorithm)
+    if isinstance(token, bytes):  # For compatibility with PyJWT 1.x
+        token = token.decode()
+    browser_integration = QCoreApplication.instance().browser_integration
+    url = 'http://%s:%s/add?token=%s' % (
+        browser_integration.host_address, browser_integration.port, token)
+    open(url)
+
+
+def _find_cluster(cluster_hash):
+    tagger = QCoreApplication.instance()
+    for cluster in tagger.clusters:
+        if hash(cluster) == cluster_hash:
+            return cluster
+    return None
+
+
+def _find_file(path):
+    tagger = QCoreApplication.instance()
+    return tagger.files.get(path, None)
+
+
+def _mbserver_url(path):
+    config = get_config()
+    host = config.setting["server_host"]
+    # port = config.setting["server_port"]
+    if host not in MUSICBRAINZ_SERVERS:
+        host = MUSICBRAINZ_SERVERS[0]  # Submission only works to official servers
```

Comment:
As we discussed in chat I will address this in a separate PR for all data submission. This will add a warning next to the server settings when the user chooses a non-standard server (explaining that data submission will go to musicbrainz.org by default), together with a check box to do data submission to the custom server anyway. We can then use this for other submission, e.g. disc ID, as well, which would also fix PICARD-1645.